### PR TITLE
Custom xcom gcs

### PR DESCRIPTION
**Pull request by rajaths010494**
Custom XCOM to upload xcom values to the GCS bucket for the given name.
Adding ENV ` AIRFLOW__CORE__XCOM_BACKEND: astronomer.providers.google.cloud.xcom_backends.gcs.GCSXComBackend` in the `docker-compose.yaml` or where the airflow is running will enable the custom XCOM for GCS.

### CODE REVIEW

**Review by phanikumv — CHANGES_REQUESTED** [review]
Can you add an example task or a DAG which shows how to use this functionality

**Comment by rajaths010494** [comment]
> Can you add an example task or a DAG which shows how to use this functionality

I have added it [here](https://github.com/astronomer/astronomer-providers/blob/custom_xcom_gcs/astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py#L12-L16). 
Add the env and build/restart the container then all the dags which run after that xcom will be pushed to GCS automatically.
Reference can be found [here](https://www.astronomer.io/guides/custom-xcom-backends/)

**Comment by bharanidharan14** [comment]
> > Can you add an example task or a DAG which shows how to use this functionality
> 
> I have added it [here](https://github.com/astronomer/astronomer-providers/blob/custom_xcom_gcs/astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py#L12-L16). Add the env and build/restart the container then all the dags which run after that xcom will be pushed to GCS automatically. Reference can be found [here](https://www.astronomer.io/guides/custom-xcom-backends/)

Doc string can be improved, you can link the reference page, env variables can be highlighted and add more information

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        """Custom XCOM for GCS to serialize the data"""
+        value = CustomXcomGCS.write_and_upload_value(value)
+        return BaseXCom.serialize_value(value)
+
+    @staticmethod
+    def deserialize_value(result: Any) -> Any:
+        """Custom XCOM for GCS to deserialize the data"""
+        result = BaseXCom.deserialize_value(result)
+        # Check if result is string and have XCOM as prefix
+        if isinstance(result, str) and result.startswith(CustomXcomGCS.PREFIX):
```
Comment (josh-fell) [inline astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py:56]: What happens if the uploaded data was bytes? I assume if the datatype of the XCom wasn't explicitly handled in the list of datatypes in the `serialize_value()` method (e.g. integers, tuples, booleans) would be uploaded to GCS as bytes?

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import os
+import uuid
+from typing import Any, Union
+
+import pandas as pd
+from airflow.models.xcom import BaseXCom
+from airflow.providers.google.cloud.hooks.gcs import GCSHook
+
+
+class CustomXcomGCS(BaseXCom):
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: changed to `GCSXComBackend`.

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Custom XCOM persistence class extends base to support various datatypes.
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: fixed

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    Add the below line in the environment
+    AIRFLOW__CORE__XCOM_BACKEND:
+    astronomer.providers.google.cloud.xcom_backend.gcs_xcom_backend.py.CustomXcomGCS
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: fixed

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    """
+
+    PREFIX = os.getenv("PREFIX", "GCSXCOM_")
+    gcp_conn_id = os.getenv("CONNECTION_NAME", "google_cloud_default")
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: fixed

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    def serialize_value(value: Any) -> Any:  # type: ignore[override]
+        """Custom XCOM for GCS to serialize the data"""
+        value = CustomXcomGCS.write_and_upload_value(value)
+        return BaseXCom.serialize_value(value)
+
+    @staticmethod
+    def deserialize_value(result: Any) -> Any:
+        """Custom XCOM for GCS to deserialize the data"""
+        result = BaseXCom.deserialize_value(result)
+        # Check if result is string and have XCOM as prefix
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: fixed.

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import json
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: renamed it to `astronomer.providers.google.cloud.xcom_backends.gcs.GCSXComBackend`

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Custom XCOM persistence class extends base to support various datatypes.
+    Add the below line in the environment
+    AIRFLOW__CORE__XCOM_BACKEND:
+    astronomer.providers.google.cloud.xcom_backend.gcs_xcom_backend.py.CustomXcomGCS
+    """
+
+    PREFIX = os.getenv("PREFIX", "GCSXCOM_")
+    gcp_conn_id = os.getenv("CONNECTION_NAME", "google_cloud_default")
+    BUCKET_NAME = os.getenv("BUCKET_NAME", "rajath_xocm_testing")
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: changed to `XCOM_BACKEND_BUCKET_NAME`

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        return result
+
+    def orm_deserialize_value(self) -> str:
+        """
+        Deserialize amethod which is used to reconstruct ORM XCom object.
+        This method should be overridden in custom XCom backends to avoid
+        unnecessary request or other resource consuming operations when
+        creating XCom orm model.
+        """
+        return "XCOM uploaded to GCS"
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: `"XCOM is uploaded into GCS bucket: {GCSXComBackend.BUCKET_NAME}"`

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        if isinstance(result, str) and result.startswith(CustomXcomGCS.PREFIX):
+            return CustomXcomGCS.read_value(result)
+        return result
+
+    def orm_deserialize_value(self) -> str:
+        """
+        Deserialize amethod which is used to reconstruct ORM XCom object.
+        This method should be overridden in custom XCom backends to avoid
+        unnecessary request or other resource consuming operations when
+        creating XCom orm model.
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: changed

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        elif isinstance(value, dict):
+            value = json.dumps(value)
+        elif isinstance(value, pd.DataFrame):
+            value = value.to_json()
+        hook.upload(CustomXcomGCS.BUCKET_NAME, key_str, data=value)
+        return key_str
+
+    @staticmethod
+    def read_value(filename: str) -> Union[str, bytes]:
+        """Downalod the file from GCS"""
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: fixed

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        # Here we download the file from GCS
+        hook = GCSHook(gcp_conn_id=CustomXcomGCS.gcp_conn_id)
+        data = hook.download(CustomXcomGCS.BUCKET_NAME, filename)
+        return data
+
+    @staticmethod
+    def serialize_value(value: Any) -> Any:  # type: ignore[override]
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: in airflow we need to add so right now i have ignored once its added in airflow we can remove this.

File: astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py (hunk 1 above)
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: Will try to add some not handled datatypes and see the result.?

Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backend/gcs_xcom_backend.py]: xcom by default pushes as dict with key and value if the value is int or any datatype it gets json.dumps into a string.

**Comment by bharanidharan14** [comment]
@josh-fell Need your review on this PR, @rajaths010494 has addressed the review comments

**Review by phanikumv — APPROVED** [review]

File: astronomer/providers/google/cloud/xcom_backends/gcs.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    def write_and_upload_value(value: Any) -> str:
+        """Convert to string and upload to GCS"""
+        key_str = GCSXComBackend.PREFIX + str(uuid.uuid4())
+        hook = GCSHook(gcp_conn_id=GCSXComBackend.GCP_CONN_ID)
+        if isinstance(value, list):
+            value = str(value)
+        elif isinstance(value, dict):
+            value = json.dumps(value)
+        elif isinstance(value, pd.DataFrame):
+            value = value.to_json()
```
Comment (feluelle) [inline astronomer/providers/google/cloud/xcom_backends/gcs.py:33]: I think you are missing a few other cases, no? What if you pass a custom obj? Shouldn't this also have an `else:` to handle the rest?

Comment (feluelle) [reply astronomer/providers/google/cloud/xcom_backends/gcs.py:33]:
```suggestion
        if isinstance(value, pd.DataFrame):
            value = value.to_json()
        else:
            value = json.dumps(value)
```
According to the [docs](https://docs.python.org/3/library/json.html#json.dumps) `json.dumps` can be of type `Any` 🤷 I am not saying it works for any type, but we could give it a try and test for which it does not. WDYT?

Comment (josh-fell) [reply astronomer/providers/google/cloud/xcom_backends/gcs.py:33]: Same thoughts as well. What would happen if users have pickling enabled (using our SDK for example)?

File: astronomer/providers/google/cloud/xcom_backends/gcs.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Custom XCom persistence class extends base to support various datatypes.
+    Add the below line in the environment
+    AIRFLOW__CORE__XCOM_BACKEND=
+    astronomer.providers.google.cloud.xcom_backends.gcs.GCSXComBackend
+    """
+
+    PREFIX = os.getenv("PREFIX", "GCSXCOM_")
+    GCP_CONN_ID = os.getenv("CONNECTION_NAME", "google_cloud_default")
+    BUCKET_NAME = os.getenv("XCOM_BACKEND_BUCKET_NAME", "some_bucket_name")
```
Comment (josh-fell) [inline astronomer/providers/google/cloud/xcom_backends/gcs.py:21]: Either users should be required to set `XCOM_BACKEND_BUCKET_NAME` or the default value should be something more proper IMO.

**Comment by josh-fell** [comment]
Generally it seems worthwhile to align this with #645 where applicable too. Seems better to have a consistently implementation regardless of underlying backend.

File: astronomer/providers/google/cloud/xcom_backends/gcs.py (hunk 14 above)
Comment (pankajastro) [reply astronomer/providers/google/cloud/xcom_backends/gcs.py:21]: +1

**Comment by rajaths010494** [comment]
trying astro-sdk to test and pyarrow to explore more datatypes.

File: astronomer/providers/google/cloud/xcom_backends/gcs.py (hunk 13 above)
Comment (rajaths010494) [reply astronomer/providers/google/cloud/xcom_backends/gcs.py]: fixed.

**Comment by rajaths010494** [comment]
Added `# nosec` to skip bandit error message since bandit has banned pickle attaching the link [here](https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html) for reference.

**Comment by josh-fell** [comment]
@rajaths010494 It looks like there are a number of unrelated changes now. Is that expected?

**Comment by rajaths010494** [comment]
> @rajaths010494 It looks like there are a number of unrelated changes now. Is that expected?

I will rebase older commits that have been merged are my PR.

**Comment by rajaths010494** [comment]
Closing this PR and raising a new PR with only GCS XCom Backend changes only. 
Please refer to the new PR [here](https://github.com/astronomer/astronomer-providers/pull/681)

**Pull request closed without merging**